File: bench/parse.bench.py
import time
import os

__dir = os.path.abspath(os.path.dirname(__file__))

digits = "0123456789"

with open(__dir + "/parse.bench.input") as f:
   prg = f.read()

def run(do_enable):
   i = 0 
   res = 0
   enable = True
   
   def is_digit():
       return prg[i] in digits
       
   def check(s):
       return prg[i:i+len(s)] == s
       
   def match(s):
       nonlocal i
       if not check(s):
           return False
       i += len(s)
       return True
       
   def num():
       nonlocal i
       n = ""
       while is_digit():
           n += prg[i]
           i += 1
       return int(n)
       
   while i < len(prg):
       if do_enable and match("don't()"):
           enable = False
       if do_enable and match("do()"):
           enable = True
           
       if match("mul("):
           a = num()
           if not match(","): 
               continue
           b = num()
           if not check(")"):
               continue
           if enable:
               res += int(a) * int(b)
       i += 1
       
   return res

start = time.perf_counter()
print(run(False))  # 168539636
print(run(True))   # 97529391
print("elapsed: " + str(time.perf_counter() - start) + "s")

File: bench/parse.bench.js
import fs from "fs";
import path from "path";
import url from "url";

const __dir = path.dirname(url.fileURLToPath(import.meta.url));

const digits = "0123456789";
const prg = fs.readFileSync(path.join(__dir, "parse.bench.input"), "utf8");

function run(do_enable) {
  let i = 0;
  let res = 0;
  let enable = true;

  function is_digit() {
    return digits.includes(prg[i]);
  }

  function check(str) {
    return prg.slice(i, i + str.length) === str;
  }

  function match(str) {
    if (!check(str)) return false;
    i += str.length;
    return true;
  }

  function num() {
    let n = "";
    while (is_digit()) {
      n += prg[i];
      i++;
    }
    return parseInt(n);
  }

  while (i < prg.length) {
    if (do_enable && match("don't()")) enable = false;
    if (do_enable && match("do()")) enable = true;

    if (match("mul(")) {
      const a = num();
      if (!match(",")) continue;
      const b = num();
      if (!check(")")) continue;
      if (enable) res += a * b;
    }
    i++;
  }
  return res;
}

const start = process.hrtime.bigint();
console.log(run(false)); // 168539636
console.log(run(true)); // 97529391
console.log(
  `elapsed: ${(Number(process.hrtime.bigint() - start) / 1_000_000_000).toFixed(
    5
  )}s`
);
